blog: packer plugin for kubevirt (#994)
kubevirt/kubevirt.github.io: _posts/2025-08-07-Packer-Plugin.md

---
layout: post
author: Ben Oukhanov
title: Create VM images with Packer
description: This post explains how to create VM images uisng Packer.
navbar_active: Blogs
pub-date: August 07
pub-year: 2025
category: news
tags:
  [
    "kubevirt",
    "kubernetes",
    "packer",
    "plugin",
    "images"
  ]
comments: true
---

# Introduction

If you're working with [Kubernetes](http://kubernetes.io) and
[KubeVirt](https://kubevirt.io), you might want to automate the creation
of virtual machine (VM) images using Packer.

[Packer](https://packer.io) is an open source tool for automating the
creation of machine images. It allows you to create identical images for

> multiple platforms from a single source template.

Author: Technically you can have multiple sources here:

```
build {
  sources = [
    "source.kubevirt-iso.rhel",
    "source.qemu.rhel"
  ]

  provisioner "shell" {
    inline = [
      "ls -la"
    ]
  }
}
```

The same provisioner(s) will be applied to both platforms.

With the addition of a dedicated KubeVirt plugin, Packer allows you to

> create VM images seamlessly, saving time and reducing

Reviewer: Saving time by running the creation in the cluster maybe?

manual configuration.

# Prerequisites

Before you begin, make sure you have the following:

- Packer installed
- A working Kubernetes cluster with KubeVirt already set up

# Plugin Features

> Packer plugin for KubeVirt offers a variety of features that simplify

Reviewer: ```suggestion
The Packer plugin for KubeVirt offers a variety of features that simplify
```

the VM image creation process:

- **HCL Templating**: Define your infrastructure as code using HCL templates.
- **ISO Installation**: Build VM images from ISO files using the `kubevirt-iso` builder.

> - **ISO Media Files**: Easily include additional files in your installation process (e.g. `ks.cfg` or `unattend.xml`).

Reviewer: Can I include other files as well? Perhaps reword this to say it allows providing files for running automated system installs?

Author: I'll rewrite this.

- **Boot Command**: Automate the VM boot process via a VNC connection with a predefined set of commands.
- **Integrated SSH/WinRM Access**: Provision and customize VMs via SSH or WinRM.

# Plugin Components

The core component of this plugin is the `kubevirt-iso` builder. This builder
allows you to start from an ISO file and create a VM image directly
on your Kubernetes cluster.

## Builder Design

> <img src="../assets/2025-08-07-Packer-Plugin/kubevirt-iso-builder-design.jpg" alt="Design" width="500"/>

Reviewer: This design can be a little overwhelming for someone not very familiar with this topic. :)

# Plugin Installation

Packer makes it easy to install plugins automatically. To install the KubeVirt
plugin, simply add the following configuration to your Packer template:

```hcl
packer {
  required_plugins {
    kubevirt = {
      source  = "github.com/kv-infra/kubevirt"
      version = ">= 0.5.0"
    }
  }
}
```

Then, run the command:

```bash
packer init
```

Packer will automatically install the KubeVirt plugin. For more detailed guidance
on how to set up and use the KubeVirt plugin with Packer,
check out the [examples](https://github.com/kv-infra/packer-plugin-kubevirt/tree/main/examples/builder/kubevirt-iso)
provided in the plugin’s repository.

# Example Template

Below is a complete Packer template that demonstrates how to use all the major
features:

- ISO-based installation from an already existing DataVolume that stores ISO file
- Embedded `ks.cfg` configuration file to automate installation
- Sending boot commands to inject `ks.cfg` in GRUB
- SSH provisioning using Shell command (`ls -la`)
- Full integration with KubeVirt’s InstanceTypes and Preferences

> This will create a Fedora VM image within a Kubernetes cluster.

Reviewer: ```suggestion
This will create a Fedora VM image running the installation and generalization process on a KubeVirt cluster.
```

```hcl
packer {
  required_plugins {
    kubevirt = {

>       source  = "github.com/kv-infra/kubevirt"

Reviewer: Should we recommend this repo from an external org?

Reviewer: Can we wait until it was moved to the hashicorp org?

Reviewer: @dominikholler @dominikholler It is still an external repository that is not managed by the KubeVirt project.

      version = ">= 0.5.0"
    }
  }
}

variable "kube_config" {
  type    = string

>   default = "${env("KUBECONFIG")}"

Author: Yes. I've added an example usage of it.

}

source "kubevirt-iso" "fedora" {
  # Kubernetes configuration
  kube_config   = var.kube_config
  name          = "fedora-42-rand-85"
  namespace     = "ben-dev"

  # ISO configuration
  iso_volume_name = "fedora-42-x86-64-iso"

  # VM type and preferences
  disk_size          = "10Gi"
  instance_type      = "o1.medium"
  instance_type_kind = "virtualmachineclusterinstancetype" # or "virtualmachineinstancetype"
  preference         = "fedora"
  preference_kind    = "virtualmachineclusterpreference" # or "virtualmachinepreference"
  os_type            = "linux"

  # Files to include in the ISO installation
  media_files = [
    "./ks.cfg"
  ]

  # Boot process configuration
  # A set of commands to send over VNC connection
  boot_command = [
    "<up>e",                            # Modify GRUB entry
    "<down><down><end>",                # Navigate to kernel line
    " inst.ks=hd:LABEL=OEMDRV:/ks.cfg", # Set kickstart file location
    "<leftCtrlOn>x<leftCtrlOff>"        # Boot with modified command line
  ]

>   boot_wait                 = "10s"     # Time to wait after boot starts

Reviewer: Slightly off topic: Why did we add this option again? Is it not sufficient to add `<wait10>` to the list of boot commands?

Author: Yes, you don't want to send boot commands when there is no expection for it from VM side.

Reviewer: Will `boot_command = [ <wait10> ]` really send something?

  installation_wait_timeout = "15m"     # Timeout for installation to complete

  # SSH configuration
  communicator      = "ssh"
  ssh_host          = "127.0.0.1"
  ssh_local_port    = 2020
  ssh_remote_port   = 22
  ssh_username      = "user"
  ssh_password      = "root"
  ssh_wait_timeout  = "20m"
}

build {
  sources = ["source.kubevirt-iso.fedora"]

  provisioner "shell" {
    inline = [
      "ls -la"
    ]
  }
}
```

Use it as a starting point to build customized VM images.

# Conclusion

> Packer with the usage of KubeVirt plugin is a powerful tool for creating VM images

Reviewer: ```suggestion
Packer in combination with the KubeVirt plugin is a powerful tool for creating VM images
```

- Packer installed
- A working Kubernetes cluster with KubeVirt already set up

## Plugin Features

The Packer plugin for KubeVirt offers a variety of features that simplify
the VM image creation process:

- **HCL Template**: Define your infrastructure as code using HCL templates.
- **ISO Installation**: Build VM images from ISO files using the `kubevirt-iso` builder.
- **ISO Media Files**: Include additional files (e.g., configs, scripts) during installation process.
- **Boot Command**: Automate the VM boot process via a VNC connection with a predefined set of commands.
- **Integrated SSH/WinRM Access**: Provision and customize VMs via SSH or WinRM.

## Plugin Components

The core component of this plugin is the `kubevirt-iso` builder. This builder
allows you to start from an ISO file and create a VM image directly
on your Kubernetes cluster.

### Builder Design

<p align="center">
  <img src="../assets/2025-08-07-Packer-Plugin/kubevirt-iso-builder-design.jpg" alt="Design" width="1125"/>
</p>

This diagram illustrates a workflow for building a bootable volume in a
Kubernetes cluster using Packer with a KubeVirt plugin.

1. Creates a temporary VM from ISO (DataVolume).

> 2. Runs Shell or Ansible provisioners over SSH.

Author: Let's simplify it by not mentioning any connection detail.

3. Converts VM's disk into a Bootable Volume.

This bootable volume can then be reused to instantiate new VMs without
repeating the installation.

## Example Usage

An example of the Packer template demonstrates key features:

- ISO-based installation from an existing DataVolume holding the ISO file.
- Embedded Kickstart configuration file to automate the installation.
- Sending boot commands to inject `ks.cfg` in GRUB.
- SSH provisioning with a Shell provisioner.
- Full integration with InstanceTypes and Preferences.

It builds a Fedora VM image inside your Kubernetes cluster.

### Export Variable

Export your kubeconfig variable, that is also used by Packer plugin:
```bash
export KUBECONFIG=~/.kube/config
```

### Deploy ISO DataVolume

Create a DataVolume to import the Fedora ISO into your cluster’s storage:
```bash
kubectl apply -f - <<EOF
apiVersion: cdi.kubevirt.io/v1beta1
kind: DataVolume
metadata:
  name: fedora-42-x86-64-iso
  annotations:
    #
    # This annotation triggers immediate binding of the PVC,
    # speeding up provisioning.
    #
    cdi.kubevirt.io/storage.bind.immediate.requested: "true"
spec:
  source:
    http:

>       url: "https://download.fedoraproject.org/pub/fedora/linux/releases/42/Server/x86_64/iso/Fedora-Server-dvd-x86_64-42-1.1.iso"

Author: I've added a note about it, but IMO let's keep it as it is.
